File: NX1980_NX1988_NX1992_NX1996_NX2000/UGOPEN/SampleNXOpenApplications/.NET/NXOpenExamples/EX_Modl_FeatureType.vb
' ==================================================================================================
'
'            Copyright (c) 2004 UGS PLM Solutions.
'                   Unpublished - All rights reserved
'
'====================================================================================================
'File description: Example for Wrapped UFUNC (NXOpen.NET)
' 
'====================================================================================================
Option Strict Off
Imports System
Imports System.IO
Imports System.Runtime.InteropServices
Imports NXOpen
Imports NXOpen.Features
Imports NXOpen.Preferences
Imports NXOpen.UF

Module EX_Modl_FeatureType

    Dim theSession As Session
    Dim theUfSession As UFSession

    Public Const UF_feature_type As Integer = 205

    Public Sub Main()
        theSession = Session.GetSession()
        theUfSession = UFSession.GetUFSession()

        Dim fs As New FileStream("EX_Modl_FeatureType.log", FileMode.Create, FileAccess.Write)
        Dim stream As New StreamWriter(fs) ' create a stream writer 
        stream.WriteLine("--Log entry goes here--")
        stream.Flush() ' update underlying file

        If File.Exists("EX_Modl_FeatureType.prt") Then
            stream.WriteLine("Remove EX_Modl_FeatureType.prt file from <Project Folder>\\bin\\Debug !!")
            stream.WriteLine("EX_Modl_FeatureType.prt already exists. !!")
            stream.Close()
            Return
        End If

        Dim nError As Integer = Execute(stream, theUfSession)
        stream.WriteLine("End of Log File")
        stream.Close()

    End Sub

    Function Execute(ByVal stream As StreamWriter, ByVal theUfSession As NXOpen.UF.UFSession) As Integer
        Try
            Dim UFPart As NXOpen.Tag
            Dim part_name As String = "EX_Modl_FeatureType"
            Dim units As Integer = 2
            Dim name As String
            'Dim Information As String

            theUfSession.Part.[New](part_name, units, UFPart)
            stream.WriteLine("New part : {0} ", part_name, " Created")

            Dim index As Integer
            Dim num_parents, num_children As Integer
            Dim enttype As Integer
            Dim feat1, master_feature As NXOpen.Tag
            Dim parent_array(), children_array() As NXOpen.Tag
            Dim part_tag As NXOpen.Tag
            Dim feat_type As String
            feat1 = 0
            master_feature = 0
            index = 0
            enttype = UF_feature_type
            Dim block_orig() As Double = {0.0, 0.0, 0.0}
            Dim block_len() As String = {"1", "2", "3"}
            Dim blk_obj As NXOpen.Tag
            theUfSession.Modl.CreateBlock1(FeatureSigns.Nullsign, block_orig, block_len, blk_obj)
            stream.WriteLine("Successful:Modl.CreateBlock1")

            'Get the tag of the work UFPart 
            part_tag = theUfSession.Assem.AskWorkPart()
            stream.WriteLine("Successful:Assem.AskWorkPart")

            'Cycle through all the features in the UFPart 
            Do
                theUfSession.Obj.CycleObjsInPart(part_tag, enttype, feat1)
                stream.WriteLine("Successful:Obj.CycleObjsInPart")

                If (feat1 = NXOpen.Tag.Null) Then
                    Exit Do
                End If
                'Get the feature type 
                theUfSession.Modl.AskFeatType(feat1, feat_type)
                stream.WriteLine("Successful:Modl.AskFeatType")

                'Information = " Feature "+ index + "  = " + feat1 + " is of type " + feat_type + "\n";
                stream.WriteLine("Feature {0} = {1} is of type {2}", index, feat1, feat_type)
                index = index + 1

                ' If the feature is an INSTANCE, then get the master feature and its type */
                If (String.Compare(feat_type, "INSTANCE") = 0) Then
                    theUfSession.Modl.AskMaster(feat1, master_feature)
                    stream.WriteLine("Successful:Modl.AskMaster")

                    theUfSession.Modl.AskFeatType(master_feature, feat_type)
                    stream.WriteLine("Successful:Modl.AskFeatType")
                    stream.WriteLine(" master feature {0} is of type = {1} ", master_feature, feat_type)

                End If
                theUfSession.Modl.AskFeatRelatives(feat1, num_parents, parent_array, num_children, children_array)
                stream.WriteLine("Successful:Modl.AskFeatRelatives")
                stream.WriteLine("Parent array for {0} contains {1} members", feat1, num_parents)

                If (num_parents > 0) Then
                    For inx As Integer = 0 To num_parents - 1
                        theUfSession.Modl.AskFeatType(parent_array(inx), feat_type)
                        stream.WriteLine("Successful:Modl.AskFeatType")
                        stream.WriteLine("  parent id {0} = {1} is of feature type {2}", inx, parent_array(inx), feat_type)
                    Next
                End If

                stream.WriteLine("  children array for feature {0} conatins {1} members", feat1, num_children)

                If (num_children > 0) Then
                    For iny As Integer = 0 To num_children - 1
                        theUfSession.Modl.AskFeatType(children_array(iny), feat_type)
                        stream.WriteLine("Successful:Modl.AskFeatType")
                        stream.WriteLine("    child id {0} = {1} is of feature type = {2}", iny, children_array(iny), feat_type)
                    Next
                End If
            Loop While (feat1 <> 0)

            'Save the Part
            theUfSession.Part.Save()
        Catch ex As Exception
            stream.WriteLine("EXCEPTION: ", ex.Message)
            Return 1
        End Try
        Return 0

    End Function

End Module

File: NX1980_NX1988_NX1992_NX1996_NX2000/UGOPEN/SampleNXOpenApplications/.NET/NXOpenExamples/EX_Modl_FeatureType.cs
/*=============================================================================

                    Copyright (c) 2004 UGS PLM Solutions
                    Unpublished - All rights reserved

===============================================================================
File description: Example for Wrapped UFUNC (NXOpen.NET)

    This example program requires a work part with at least one feature. The
    example extracts the feature information of a given solid body via 
    theUfSession.Modl.AskFeatType(), theUfSession.Modl.AskMaster(), 
    theUfSession.Modl.AskFeatRelatives() etc. The program  prints a report 
    to the listing window.
 */
using System;
using System.IO;
using NXOpen;
using NXOpen.UF;

namespace NetExample
{
    /// This example program requires a work part with at least one feature. 
    /// The example extracts the feature information of a given solid body via theUfSession.Modl.AskFeatType(),
    /// theUfSession.Modl.AskMaster(), theUfSession.Modl.AskFeatRelatives() etc. 
    /// The program  prints a report to the listing window.
    public class EX_Modl_FeatureType
    {
        public static UFSession theUfSession;
        private static Session theSession;
        
        public int Execute()
         {
             
             int index, inx;
             int num_parents, num_children;
             int enttype;
             Tag feat1, master_feature;
             Tag[] parent_array, children_array;
             Tag part_tag;
             string feat_type = null;
             string tmp_text = null;
             feat1 = 0;
             master_feature = 0;
             index = 0;
             enttype = UFConstants.UF_feature_type; //UF_feature_type;
             double[] block_orig = {0.0,0.0,0.0};
             string[] block_len = {"1","2","3"};
            
             /* Get the tag of the work UFPart */
             part_tag = theUfSession.Assem.AskWorkPart();
             /* Cycle through all the features in the UFPart */
             do
             {
                 theUfSession.Obj.CycleObjsInPart(part_tag, enttype, ref feat1);
                 if(feat1 == Tag.Null) break;
                 /* Get the feature type */
                 theUfSession.Modl.AskFeatType(feat1,out feat_type);
                 tmp_text = " Feature "+ index + "  = " + feat1 + " is of type " + feat_type + "\n";
                 theUfSession.Ui.WriteListingWindow(tmp_text);
                 
                 ++index;
                 /* If the feature is an INSTANCE, then get the master feature and its type */
                 if (String.Compare(feat_type,"INSTANCE") == 0)
                 {
                     theUfSession.Modl.AskMaster(feat1,out master_feature);
                     theUfSession.Modl.AskFeatType(master_feature,out feat_type);
                     tmp_text = " Master Feature "+ master_feature + " is of type " + feat_type + "\n";
                     theUfSession.Ui.WriteListingWindow(tmp_text);
                 }
                 theUfSession.Modl.AskFeatRelatives(feat1,out num_parents,out parent_array,out num_children,
                     out children_array);

                 tmp_text =  "  Parent array for " + feat1 + " contains " + num_parents + " members:\n";
                 theUfSession.Ui.WriteListingWindow(tmp_text);
                 if(num_parents > 0)
                 {
                     for (inx = 0;inx < num_parents; inx++)
                     {
                         theUfSession.Modl.AskFeatType(parent_array[inx],out feat_type);
                         tmp_text =  "  Parent id " + inx + " = " + parent_array[inx] +
                                     " is of feature type " + feat_type + "\n";
                         theUfSession.Ui.WriteListingWindow(tmp_text);
                     }
                 }
                 
                 tmp_text = "  children array for feature " + feat1 + " contains " + num_children + "  members:\n";
                 theUfSession.Ui.WriteListingWindow(tmp_text);

                 if(num_children > 0)
                 {
                     for (inx = 0;inx < num_children; inx ++)
                     {
                         theUfSession.Modl.AskFeatType(children_array[inx],out feat_type);
                         tmp_text = "  child is " + inx + " = " + children_array[inx]
                                    + " is of feature type " + feat_type + "\n";
                         theUfSession.Ui.WriteListingWindow(tmp_text);
                     }
                 }

             }while(feat1 != 0);
             
            return 0;
        }

        public static void Main(string[] args)
        {
            string tmp_text = null;
            theSession=Session.GetSession();
            theUfSession= UFSession.GetUFSession();

            /* Open Listing Window */
            theUfSession.Ui.OpenListingWindow();

            theUfSession.Ui.WriteListingWindow("Log Entry : \r\n");
            theUfSession.Ui.WriteListingWindow("--Log entry goes here--\n");
            
            try
            {
                EX_Modl_FeatureType curveTest1 = new EX_Modl_FeatureType();
                if (curveTest1.Execute()==0)
                {
                    theUfSession.Ui.WriteListingWindow("Successful\n");
                }
                else
                {
                    theUfSession.Ui.WriteListingWindow("Failed\n");
                }
            }
            catch(NXException e)
            {   
                tmp_text = "Exception is: " + e.Message + "\n";
                theUfSession.Ui.WriteListingWindow(tmp_text);
            }
            
            theUfSession.Ui.WriteListingWindow("End of Log File\n");
            //theUfSession.Ui.CloseListingWindow();
        }

        public static int GetUnloadOption(string dummy) 
        {
            return UFConstants.UF_UNLOAD_IMMEDIATELY;
        }
    }
}
